### PR TITLE
fix(input-field): sync invalid state to underlying mdctextfield

### LINKED ISSUE #791: input-field not respecting invalid prop
## Current behavior

The mdcTextField will not respect the invalid class put onto the containing label component and will instead reset the classes based on its own calculation of whether or not the component is valid, likely based on native input attributes such as required and the type.

### Steps to reproduce
* Pull master
* Get lime-elements into the browser somehow
* Insert `<limel-text-field invalid help-text="help text" value=""></limel-input-field>` into the dom
* The text field should be red signifying that it is invalid (since the invalid prop is set). 
* Click to focus the input field
* Click away to blur the input field
* Note that the red styling due to the invalid state is missing due to loss of focus. This is incorrect the red styling should remain since the invalid prop is set

## Expected behavior

If I set the invalid prop to true the field should be shown as invalid regardless of the internal validation of the text field.

## Environment

- **lime-elements** version: 26.27.0
- Framework used: N/A the issue shows when using lime-elements in an empty html page.
- Logs: N/A

## Proposed Solutions
https://github.com/Lundalogik/lime-elements/pull/788

### PR DESCRIPTION
**Pull request by sampsonbryce**
The mdcTextField will not respect the invalid class put onto the containing label component
and will instead reset the classes based on its own calculation of whether or not the component
is valid, likely based on native input attributes such as required and the type. This sync update
forces the mdcTextField into an invalid state if the invalid prop is set, otherwise it
lets mdcTextField determine if the component is valid or not.

## Steps to reproduce
* Pull master
* Get lime-elements into the browser somehow
* Insert `<limel-text-field invalid help-text="help text" value=""></limel-input-field>` into the dom
* The text field should be red signifying that it is invalid (since the invalid prop is set). 
* Click to focus the input field
* Click away to blur the input field
* Note that the red styling due to the invalid state is missing due to loss of focus. This is incorrect the red styling should remain since the invalid prop is set

fix #791

## Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/input-field/input-field.tsx [hunk 1]
```diff
@@ -570,4 +574,17 @@ export class InputField {
...
         }
     }
+
+    private syncInvalidToTextField() {
+        if (!this.mdcTextField) {
+            return;
+        }
+
+        if (!this.mdcTextField.valid) {
+            return;
```
Comment (adrianschmidt) [inline src/components/input-field/input-field.tsx:584]: > This sync update forces the mdcTextField into an invalid state if the invalid prop is set, otherwise it lets mdcTextField determine if the component is valid or not.

Hmm… 🤔

What happens if, for some reason, the consumer's validation says the input is valid, but the MDC component says it's not? The component would render as invalid, I assume, but the value of limel-input-field's `invalid` property would be `false, right? That might mean that a Save-button that normally becomes disabled when there are validation errors will be enabled.

Regardless, that should reasonably be a much less severe bug than the bug this fixes, but we might want to try it out so we know.

File: src/components/input-field/input-field.tsx [hunk 2]
```diff
@@ -570,4 +574,17 @@ export class InputField {
...
+        }
+
+        const isValid = !this.isInvalid();
```
Comment (adrianschmidt) [inline src/components/input-field/input-field.tsx:587]: The `isInvalid` function looks a little weird to me, especially when `syncInvalidToTextField` is added to the mix.

Seems like we're now juggling the `invalid` property set by the consumer, the validity reported by the *native* element, and the validity reported by the MDC component, and it's spread out over two different functions where one is calling the other, but both are called by other code at different points in the life cycle. I think it's too complex and inconsistent. If we really need to consider all these three values, we need to do it consistently, in a single place, and make sure all those values are considered every time any of them are updated.

File: src/components/input-field/input-field.tsx (hunk 1 above)
Comment (sampsonbryce) [reply src/components/input-field/input-field.tsx:584]: Ya im not 100% sure how we could make the save button disabled if the MDC internal validation fails. From what i understand of this code the MDC internal validation is not exported in any way so theres no way to access it outside of limel-input-field right?

File: src/components/input-field/input-field.tsx (hunk 2 above)
Comment (sampsonbryce) [reply src/components/input-field/input-field.tsx:587]: Ya it is kinda hard to follow. I'm not 100% sure the best way to go about this. If you want I can port this PR to an issue and link this as a potential solution so that the best way to fix this can be discussed outside of the context of a specific implementation.

Comment (jgroth) [reply src/components/input-field/input-field.tsx:587]: I'm not 100% sure, but I think there is only the `invalid` property and the valid status of the native element. Material is using the native validity reported as well if I remember correctly. The `invalid` property should probably always have precedence over anything reported by the element

Comment (adrianschmidt) [reply src/components/input-field/input-field.tsx:587]: > If you want I can port this PR to an issue and link this as a potential solution

Yeah, that would be great. It's generally a good idea to create a bug report first, even if you already have a PR ready to go (not that I always do it myself 😖).

Comment (sampsonbryce) [reply src/components/input-field/input-field.tsx:587]: issue: https://github.com/Lundalogik/lime-elements/issues/791

**Comment by BregenzerK** [comment]
I looked into the issue (because I also didn't like the sync method too much). I think we have two cases that we want to support: 
- the native validation depending on the type and properties like `minlength`
- a consumer validation, that ends up defining the `invalid` property based on consumer logic that the native validation check can't take into consideration

In my opinion we just need to overwrite the native valid state if the consumer 'knows better' to slso support the second case. I also adjust the example to show both cases.

What do you think?
@adrianschmidt @jgroth

**Comment by adrianschmidt** [comment]
> In my opinion we just need to overwrite the native valid state if the consumer 'knows better' to also support the second case. I also adjust the example to show both cases.

Yeah, that sounds good to me 👍

File: src/components/input-field/examples/input-field-error-icon.tsx [hunk 3]
```diff
@@ -7,25 +7,49 @@ const MIN_LENGTH = 6;
...
+        this.valueConsumer = event.detail;
+    }
+
+    private isInvalid() {
+        const substringLength = 9;
+        return !(
+            this.valueConsumer &&
+            this.valueConsumer.substr(-substringLength, substringLength) ===
+                '@test.com'
+        );
```
Comment (adrianschmidt) [inline src/components/input-field/examples/input-field-error-icon.tsx:53]: I added this small improvement of the example validator.

If you think it makes the example harder to understand, feel free to veto it :)

File: src/components/input-field/input-field.tsx [hunk 4]
```diff
@@ -188,6 +188,12 @@ export class InputField {
         }
     }
 
+    public componentDidUpdate() {
+        if (this.invalid) {
+            this.mdcTextField.valid = false;
```
Comment (adrianschmidt) [inline src/components/input-field/input-field.tsx:193]: Wow, nice simplification! 👏

File: src/components/input-field/examples/input-field-error-icon.tsx (hunk 3 above)
Comment (BregenzerK) [reply src/components/input-field/examples/input-field-error-icon.tsx:53]: looks fine, I wasn't going for absolute perfection on the validator, just wanted to make a point ;) but if that is exacter then go with it
